### DISCUSSION TITLE
compare fails with strange error on png file

### DISCUSSION THREAD

**Discussion by travisdowns** [ImageMagick/ImageMagick#4352] [Development]
Consider [this rather tall png file](https://user-images.githubusercontent.com/2403521/136885741-850cd560-2852-49ac-8e10-9f81e79e64b7.png).

If you download it as `a.png` and run 

    compare a.png a.png null:
    
You get the following error:

~~~
compare-im6.q16: missing an image filename `a.png' @ error/compare.c/CompareImageCommand/1151).
~~~

Not other useful output and neither `strace` nor `ltrace` are very illuminating (`ltrace` in particular doesn't show anything for IM on Ubuntu).

    Version: ImageMagick 6.9.10-23 Q16 x86_64 20190101

**Comment by fmw42** [comment]
Compare requires an output. If you do not want it, use null:

`compare a.png a.png null:`

**Reply by travisdowns** [reply]
> Compare requires an output. If you do not want it, use null:

It has an output in my example: `/dev/null`

The behavior is the same with `null:` instead. 

It's related to the files somehow. The same command line with another png works fine.

I updated the example command line to use the more idiomatic `null:` to avoid any confusion.

**Reply by fmw42** [reply]
Works fine for me on IM 6.9.12-24 Q16 Mac OSX Sierra.  What is your IM version and platform?  Have you tried use -metric rmse rather than the default NCC?

**Reply by travisdowns** [reply]
IM version is `Version: ImageMagick 6.9.10-23 Q16 x86_64 20190101`.

~~~
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.3 LTS
Release:	20.04
Codename:	focal
~~~

`-metric rmse` produces the same result. I thing the failure is earlier: based on the `strace` call, it occurs immediately after `mmaping` the file: it is immediately closed and the error output without any intervening system calls.

**Reply by travisdowns** [reply]
FWIW, I just built `ImageMagick 7.1.0-10` from source and the error goes away. So it was either fixed in a later version or something about building it from source makes it go away.  Oh well.

**Comment by snibgo** [comment]
With IM v6.9.9-50, the command...
```
compare -metric RMSE tallpng.png tallpng.png NULL:
```
... works fine. Note that a "-metric" should always be given. The file is fairly large, at 21 million pixels. On a small computer, this may cause problems.

**Reply by travisdowns** [reply]
> Note that a "-metric" should always be given.

Well if you don't give it, you'll use NCC by default, right?



> he file is fairly large, at 21 million pixels. On a small computer, this may cause problems.

Sure, but I don't think that's the cause here. This machine has 16 GB of RAM and plenty of free disk space. With a version that doesn't crash the comparison takes only a few seconds. The crash is immediate also, not after doing a bunch of work indicating an "early" failure.

> With IM v6.9.9-50, the command... ... works fine.

Well it's not just me, probably any Ubuntu 20.04 user. Here's the same problem reproduced on github:

https://github.com/travisdowns/gha-test/runs/3888836434?check_suite_focus=true#step:4:1

**Comment by snibgo** [comment]
> Well if you don't give it, you'll use NCC by default, right?

No. For example with v7.1.0-4 on Windows 8.1, with no metric:
```
f:\web\im>%IMG7%magick compare toes.png toes.png NULL:

```
Compare to "-metric NCC":
```
f:\web\im>%IMG7%magick compare -metric NCC toes.png toes.png NULL:
0.999984
```
See the difference? The first command gave no text output.

I don't know what is causing your problem. Perhaps it is specific to Ubuntu installations.
